### **Grupo 7**
*  Mara Alejandra Calderon Díaz
*  Magalys del Carmen Camargo Hernandez
*  Luis Eduardo Osorio Acevedo

# **Guía 2**

Objetivo: aplicar los conceptos de Pandas en la manipulación y análisis de datos estructurados. Se trabajará con Series y DataFrames, aplicando técnicas de acceso, selección, filtrado y agregación de datos.

### **Caso de Negocio: Análisis de Deserción de Clientes en una Empresa de Telecomunicaciones**

**Contexto**

En la industria de las telecomunicaciones, la retención de clientes es un factor crítico para la sostenibilidad del negocio. La competencia es feroz y adquirir nuevos clientes suele ser más costoso que mantener a los actuales. En este análisis, exploraremos un conjunto de datos que contiene información detallada sobre clientes que han abandonado el servicio (churn) y aquellos que permanecen activos.

Nuestro objetivo es identificar los factores clave que influyen en la deserción, comprender patrones de comportamiento y generar estrategias efectivas para reducir la tasa de abandono.

La empresa ha experimentado un aumento en la tasa de deserción de clientes, lo que ha provocado:
- Pérdida de ingresos recurrentes.
- Incremento en los costos de adquisición de nuevos clientes.
- Menor estabilidad en la base de clientes a largo plazo.

Para abordar esta problemática, es crucial identificar las razones detrás de la deserción y desarrollar estrategias para mejorar la retención de clientes.



**Entrega del Trabajo**

Los estudiantes deben trabajar en grupos de entre 2 y 4 personas. Cada grupo deberá subir su trabajo a un repositorio de GitHub, asegurándose de que el código y los archivos necesarios estén bien organizados y documentados. Posteriormente, deberán enviar el enlace del repositorio en la plataforma Canvas para su evaluación.

Instrucciones para la entrega:

Crear un repositorio en GitHub con un nombre descriptivo para el proyecto.

Subir el código en Jupyter Notebook (.ipynb) o en formato Python (.py).

Incluir un archivo README.md con una breve descripción del trabajo y las instrucciones de ejecución.

Compartir el enlace del repositorio en Canvas dentro del plazo establecido.



**Descripción de las variables del dataset telecom_churn**

El dataset telecom_churn contiene información detallada sobre clientes de una empresa de telecomunicaciones, incluyendo datos generales, planes contratados, uso del servicio telefónico y llamadas al servicio al cliente. Su propósito principal es analizar patrones de deserción de clientes, identificados a través de la variable churn, que indica si un cliente ha abandonado la empresa (1) o sigue siendo cliente (0).

Dentro del dataset, encontramos información general como el estado (state) donde reside el cliente, el código de área (area code), y el número de teléfono (phone number), aunque esta última variable no aporta información útil para el análisis, ya que es un identificador único.

Además, el dataset registra el tiempo que un cliente ha estado en la empresa a través de la variable account length, lo que puede ayudar a analizar si la duración del contrato influye en la deserción. También se incluyen detalles sobre los planes contratados, como si el cliente tiene un plan internacional (international plan), que le permite realizar llamadas internacionales, o un buzón de voz (voice mail plan), que le permite recibir mensajes de voz.

En cuanto al uso del servicio, se registran datos detallados sobre el tiempo en llamadas y los costos asociados. Se divide en tres períodos del día: diurno (total day minutes, total day calls, total day charge), vespertino (total eve minutes, total eve calls, total eve charge) y nocturno (total night minutes, total night calls, total night charge), lo que permite evaluar si hay patrones de consumo que influyen en la deserción. También se incluye información sobre el uso del servicio internacional, con variables como total intl minutes (minutos en llamadas internacionales), total intl calls (cantidad de llamadas internacionales) y total intl charge (costos por llamadas internacionales).

Otro aspecto clave del dataset es el número de llamadas al servicio al cliente (customer service calls), ya que una mayor cantidad de llamadas puede indicar insatisfacción y estar relacionada con la decisión del cliente de abandonar la empresa.

Finalmente, la variable más importante del análisis es churn, que indica si un cliente ha desertado de la empresa. A partir de esta variable, podemos analizar qué factores influyen en la deserción y encontrar patrones en los clientes que tienen mayor probabilidad de abandonar el servicio.


**Exploración y Limpieza de Datos**

Cargar y explorar el dataset

Importa Pandas y carga el dataset telecom_churn.csv en un DataFrame.

Muestra las primeras 5 filas del DataFrame.

Verifica cuántas filas y columnas tiene el dataset.

Muestra información general del dataset, incluyendo los tipos de datos.

Identifica si hay valores nulos en alguna columna.

In [3]:
# Se importa la librería pandas
import pandas as pd

# Se carga el dataset telecom_churn.csv en un DataFrame
df = pd.read_csv('..Dataset/telecom_churn.csv')

# Se muestran las primeras 5 filas del DataFrame
df.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

In [4]:
# Se verifica la cantidad de filas y columnas del dataset
df.shape

(3333, 21)

In [5]:
# Se muestra información general del dataset, incluyendo tipos de datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3333 non-null   object 
 1   account length          3333 non-null   int64  
 2   area code               3333 non-null   int64  
 3   phone number            3333 non-null   object 
 4   international plan      3333 non-null   object 
 5   voice mail plan         3333 non-null   object 
 6   number vmail messages   3333 non-null   int64  
 7   total day minutes       3333 non-null   float64
 8   total day calls         3333 non-null   int64  
 9   total day charge        3333 non-null   float64
 10  total eve minutes       3333 non-null   float64
 11  total eve calls         3333 non-null   int64  
 12  total eve charge        3333 non-null   float64
 13  total night minutes     3333 non-null   float64
 14  total night calls       3333 non-null   

Según la información obetenida el data set cuenta con 3333 registros y 21 columnas. De estas columnas 1 es un tipo de datos boleano, 8 flotantes, 8 enteros y 4 objetos.

**Análisis de Churn y Factores Relacionados**

Calcula el porcentaje de clientes que han desertado (churn = 1).

Identifica si los clientes con plan internacional (international plan) tienen mayor tasa de deserción.

Identifica si los clientes con buzón de voz (voice mail plan) tienen menor tasa de deserción.

In [7]:
# Porcentaje de clientes que han desertado
client_desercion = df['churn'].value_counts(normalize= True)*100
client_desercion

churn
False    85.508551
True     14.491449
Name: proportion, dtype: float64

In [8]:
# Agrupar datos por plan internacional
df.groupby(['international plan', 'churn']).size().get("yes")

churn
False    186
True     137
dtype: int64

In [9]:
# Agrupar datos por plan de buzon de voz
df.groupby(['voice mail plan', 'churn']).size().get('yes')

churn
False    842
True      80
dtype: int64

Conclusión: No se puede afirmar que los clientes con plan de buzon de voz tienen una mayor taza de deserción

**Análisis de la Duración del Servicio y Deserción**

¿Cuál es la duración promedio de la cuenta (account length) entre clientes que desertaron y los que permanecen?

¿Los clientes con cuentas más antiguas tienen más probabilidades de desertar?

In [10]:
# Sacar duración promedio
df.groupby('churn')["account length"].mean()

churn
False    100.793684
True     102.664596
Name: account length, dtype: float64

In [11]:
# Sacar promedio de la variable a account length
prom = df['account length'].mean()

# Filtrar base de datos extrayendo clientes con cuentas con una duración superior al promedio
df_cuent_ant = df.loc[df['account length'] >= prom]

# Contar desertores y permanentes en data set df_cuent_ant
print("Desertores y permanentes df_cuent_ant\n",df_cuent_ant['churn'].value_counts())

Desertores y permanentes df_cuent_ant
 churn
False    1385
True      244
Name: count, dtype: int64


**Relación entre Deserción y Uso del Servicio**

Compara la cantidad de minutos usados en llamadas diurnas (total day minutes) entre clientes con y sin churn.

Compara la cantidad de minutos usados en llamadas nocturnas (total night minutes).

Compara el número total de llamadas (total day calls) entre clientes con y sin churn.

In [12]:
# Sumar total day minutes por cada categoria de la variable churn
print("Total day minutes por cada categoria de la variable churn\n",df.groupby('churn')["total day minutes"].sum())
# Sumar total night minutes por cada categoria de la variable churn
print("\nTotal night minutes por cada categoria de la variable churn\n",df.groupby('churn')["total night minutes"].sum())
# Sumar total day calls por cada categoria de la variable churn
print("\nTotal day calls por cada categoria de la variable churn\n",df.groupby('churn')["total day calls"].sum())

Total day minutes por cada categoria de la variable churn
 churn
False    499250.9
True      99939.5
Name: total day minutes, dtype: float64

Total night minutes por cada categoria de la variable churn
 churn
False    570379.6
True      99126.9
Name: total night minutes, dtype: float64

Total day calls por cada categoria de la variable churn
 churn
False    285807
True      48945
Name: total day calls, dtype: int64


**Impacto de las Llamadas al Servicio al Cliente en la Deserción**

Calcula el número promedio de llamadas al servicio al cliente (customer service calls) entre clientes que desertaron y los que no.

Divide los clientes en dos grupos:

- Grupo 1: Clientes que llamaron más de 3 veces al servicio al cliente.
- Grupo 2: Clientes que llamaron 3 veces o menos.
Compara la tasa de churn entre ambos grupos.

In [13]:
# Promedio de llamadas al servicio al cliente por cada categoria de la variable churn
print("Promedio de llamadas al servicio al cliente entre clientes que desertaron y los que no\n",df.groupby('churn')["customer service calls"].mean())

Promedio de llamadas al servicio al cliente entre clientes que desertaron y los que no
 churn
False    1.449825
True     2.229814
Name: customer service calls, dtype: float64


In [14]:
# Declarar variables
cand_llam = 3
# Filtrar base de datos extrayendo clientes que llamaron más de 3 veces al servicio al cliente
df_cand_llam_mas = df.loc[df['customer service calls'] > cand_llam]
# Filtrar base de datos extrayendo clientes que llamaron 3 veces o menos al servicio al cliente
df_cuent_llam_men = df.loc[df['customer service calls'] <= cand_llam]
print(df_cand_llam_mas.shape)
print(df_cuent_llam_men.shape)
# Contar desertores y permanentes en data set df_cand_llam_mas
print("Tasa desertores y permanentes df_cand_llam_mas\n",df_cand_llam_mas['churn'].value_counts(normalize= True)*100)
# Contar desertores y permanentes en data set df_cuent_llam_men
print("Tasa desertores y permanentes df_cuent_llam_men\n",df_cuent_llam_men['churn'].value_counts(normalize= True)*100)

(267, 21)
(3066, 21)
Tasa desertores y permanentes df_cand_llam_mas
 churn
True     51.685393
False    48.314607
Name: proportion, dtype: float64
Tasa desertores y permanentes df_cuent_llam_men
 churn
False    88.747554
True     11.252446
Name: proportion, dtype: float64


**Análisis del Costo de las Llamadas y Churn**

Compara el costo total de llamadas diurnas (total day charge) entre clientes con y sin churn.

Compara el costo total de llamadas nocturnas (total night charge).

¿Los clientes con mayor gasto en llamadas internacionales (total intl charge) tienen más probabilidades de desertar?

In [16]:
# Comparacion el costo total de llamadas entre clientes con y sin churn

agrupado = df.groupby("churn")["total day charge"].sum()
print(agrupado)

if False>True:
  print("La tasa de churn entre los clientes con churn es mayor")
else:
  print("La tasa de churn entre los clientes sin churn es mayor")

# Comparacion del costo total de llamadas nocturnas

agrupado = df.groupby("churn")["total night charge"].sum()
print("\nEl costo total de llamadas nocturnas entre clientes con y sin churn")
print(agrupado)

churn
False    84874.20
True     16989.97
Name: total day charge, dtype: float64
La tasa de churn entre los clientes sin churn es mayor

El costo total de llamadas nocturnas entre clientes con y sin churn
churn
False    25667.31
True      4460.76
Name: total night charge, dtype: float64
